### PR TITLE
Additional QBFT integration tests

### PR DESCRIPTION
**Pull request by jframe**
Adds additional QBFT integration tests for the QBFT state machine. Leaving out the validation integration tests until the message validation is complete.

### CODE REVIEW

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/GossipTest.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+  @Test
+  public void futureMessageGetGossipedLater() {
+    final Block signedCurrentHeightBlock =
+        BftHelpers.createSealedBlock(block, peers.sign(block.getHash()));
+
+    ConsensusRoundIdentifier futureRoundId = new ConsensusRoundIdentifier(2, 0);
+    Prepare futurePrepare = sender.injectPrepare(futureRoundId, block.getHash());
+    peers.verifyNoMessagesReceivedNonProposing();
+    ;
```
Comment (usmansaleem) [inline consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/GossipTest.java:169]: nit: additional ;

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/GossipTest.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    sender.injectProposal(futureRoundId, block);
+    peers.verifyNoMessagesReceived();
+  }
+
+  @Test
+  public void futureMessageGetGossipedLater() {
+    final Block signedCurrentHeightBlock =
+        BftHelpers.createSealedBlock(block, peers.sign(block.getHash()));
+
+    ConsensusRoundIdentifier futureRoundId = new ConsensusRoundIdentifier(2, 0);
```
Comment (usmansaleem) [inline consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/GossipTest.java:166]: nit: final

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/LocalNodeNotProposerTest.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    peers.commit(roundId, blockToPropose.getHash());
+    peers.verifyNoMessagesReceived();
+    assertThat(context.getCurrentChainHeight()).isEqualTo(0);
+
+    peers.prepareForNonProposing(roundId, blockToPropose.getHash());
+    peers.verifyNoMessagesReceived();
+    assertThat(context.getCurrentChainHeight()).isEqualTo(0);
+
+    peers.getProposer().injectProposal(roundId, blockToPropose);
+    // TODO(tmm): Unfortunately, there are times that the Commit will go out BEFORE the prepare
```
Comment (usmansaleem) [inline consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/LocalNodeNotProposerTest.java:109]: nit: TODO meant to be removed?

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/ReceivedFutureProposalTest.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+import org.hyperledger.besu.ethereum.core.Block;
+
+import java.util.Collections;
+import java.util.List;
+
+import org.junit.Ignore;
+import org.junit.Test;
+
+/**
+ * Ensure the Ibft component responds appropriately when a future round Proposal message is
```
Comment (usmansaleem) [inline consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/ReceivedFutureProposalTest.java:40]: nit: doco

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/round/QbftRoundIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    when(messageValidator.validateProposal(any())).thenReturn(true);
+    when(messageValidator.validatePrepare(any())).thenReturn(true);
+    when(messageValidator.validateCommit(any())).thenReturn(true);
+
+    when(nodeKey.sign(any())).thenThrow(new SecurityModuleException("Hsm Is Down"));
+
+    throwingMessageFactory = new MessageFactory(nodeKey);
+    transmitter = new QbftMessageTransmitter(throwingMessageFactory, multicaster);
+
+    BftExtraData proposedExtraData =
```
Comment (usmansaleem) [inline consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/round/QbftRoundIntegrationTest.java:100]: nit: final

**Review by usmansaleem — APPROVED** [review]
LGTM - minor cosmetic enhancements requested.

**Review by usmansaleem — APPROVED** [review]
LGTM

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/round/QbftRoundIntegrationTest.java (hunk 5 above)
Comment (jframe) [reply consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/round/QbftRoundIntegrationTest.java:100]: done

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/ReceivedFutureProposalTest.java (hunk 4 above)
Comment (jframe) [reply consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/ReceivedFutureProposalTest.java]: done

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/LocalNodeNotProposerTest.java (hunk 3 above)
Comment (jframe) [reply consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/LocalNodeNotProposerTest.java:109]: no unfortunately. this hasn't been fixed yet

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/GossipTest.java (hunk 2 above)
Comment (jframe) [reply consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/GossipTest.java]: done

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/GossipTest.java (hunk 1 above)
Comment (jframe) [reply consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/GossipTest.java]: done

**Review by usmansaleem — APPROVED** [review]
LGTM

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/support/TestContextBuilder.java [hunk 6]
```diff
@@ -161,7 +161,7 @@ public TestContextBuilder clock(final Clock clock) {
     return this;
   }
 
-  public TestContextBuilder ibftEventQueue(final BftEventQueue bftEventQueue) {
+  public TestContextBuilder bftEventQueue(final BftEventQueue bftEventQueue) {
```
Comment (rain-on) [inline consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/support/TestContextBuilder.java:164]: why not just "eventQueue"?

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/FutureHeightTest.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+  @Test
+  public void multipleNewChainHeadEventsDoesNotRestartCurrentHeightManager() {
+    final Block currentHeightBlock = context.createBlockForProposalFromChainHead(0, 30);
+
+    peers.getProposer().injectProposal(roundId, currentHeightBlock);
+    peers.getNonProposing(0).injectPrepare(roundId, currentHeightBlock.getHash());
+
+    peers.clearReceivedMessages();
+
+    // inject a NewHeight FOR THE CURRENT HEIGHT
```
Comment (rain-on) [inline consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/FutureHeightTest.java:147]: comment is meaningless, code says this already

**Review by rain-on — APPROVED** [review]
LGTM ...

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/support/TestContextBuilder.java (hunk 6 above)
Comment (jframe) [reply consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/support/TestContextBuilder.java:164]: true, don't need the distinction. done.

File: consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/FutureHeightTest.java (hunk 7 above)
Comment (jframe) [reply consensus/qbft/src/integration-test/java/org/hyperledger/besu/consensus/qbft/test/FutureHeightTest.java:147]: done

**Pull request merged**